### PR TITLE
Support awsQueryCompatible trait: Parse error code when present in HTTP response header

### PR DESCRIPTION
**Pull request by haocheng-aws**
Support awsQueryCompatible trait: Parse error code when present in HTTP response header.

## Motivation

As we decided to parse AwsQuery Compatible Error in the Http response header, the required changes need to take in place on both SDK client and Service.
This will be a pre-requisite for migrating services from AwsQuery wire protocol to AwsJson.

## What changed

Error code can be fetched from response content (for example `__type#fooException`). This PR takes another scenario into consideration, of which when `x-amzn-query-error` header with valid value is found from the Http response.

The parsed error code remains unchanged if this header is missing, or contains null or invalid value.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/json/error_handler.rb [hunk 1]
```diff
@@ -26,7 +26,11 @@ def extract_error(body, context)
       end
 
       def error_code(json, context)
-        code = json['__type']
+        code = if context.config.api.metadata['awsQueryCompatible'] and context.http_response.headers['x-amzn-query-error']
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/json/error_handler.rb]: I think we should wrap this check in a "aws_query_error?(context)" method.

File: gems/aws-sdk-core/spec/api_helper.rb [hunk 2]
```diff
@@ -13,6 +13,14 @@ def sample_json # dynamodb
       end
     end
 
+    def sample_json_aws_query_compatible # sqs
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/api_helper.rb]: I think we should remove this - it won't get much re-use. I would prefer if, in error_handler_spec, we define a minimal API. There are some examples (endpoint_discovery_spec or all in that plugins folder).

File: gems/aws-sdk-core/spec/aws/json/error_handler_spec.rb [hunk 3]
```diff
@@ -98,6 +98,45 @@ module Json
...
+          stub_request(:post, "https://sqs.us-west-2.amazonaws.com/").
+            to_return(:status => 400, headers: {
+              'x-amzn-query-error': 'AwsQueryError;Sender'
+            }, :body => error_resp)
+          expect {
+            client_aws_query_compatible.send_message(
+              queue_url: 'https://queue.url',
+              message_body: 'message body'
+            )
+          }.to raise_error {|e|
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/json/error_handler_spec.rb]: For multiline ruby blocks, don't use `{}` but instead use `do end`.

File: gems/aws-sdk-core/spec/fixtures/apis/sqs.json [hunk 4]
```diff
@@ -0,0 +1,1147 @@
+{
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/fixtures/apis/sqs.json]: Can be deleted if following above recommendation

File: gems/aws-sdk-core/spec/api_helper.rb (hunk 2 above)
Comment (haocheng-aws) [reply gems/aws-sdk-core/spec/api_helper.rb]: Thanks a lot for advice! Updated.

**Review by mullermp — APPROVED** [review]
Nice!

**Review by alextwoods — APPROVED** [review]

**Pull request merged**